core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1108,40 +1162,29 @@ private[spark] class TaskSetManager(
-      // bound based on that.
-      logDebug("Task length threshold for speculation: " + threshold)
-      for (tid <- runningTasksSet) {
-        var speculated = checkAndSubmitSpeculatableTask(tid, time, threshold)
-        if (!speculated && executorDecommissionKillInterval.isDefined) {
-          val taskInfo = taskInfos(tid)
-          val decomState = sched.getExecutorDecommissionState(taskInfo.executorId)
-          if (decomState.isDefined) {
-            // Check if this task might finish after this executor is decommissioned.
-            // We estimate the task's finish time by using the median task duration.
-            // Whereas the time when the executor might be decommissioned is estimated using the
-            // config executorDecommissionKillInterval. If the task is going to finish after
-            // decommissioning, then we will eagerly speculate the task.
-            val taskEndTimeBasedOnMedianDuration = taskInfos(tid).launchTime + medianDuration
-            val executorDecomTime = decomState.get.startTime + executorDecommissionKillInterval.get
-            val canExceedDeadline = executorDecomTime < taskEndTimeBasedOnMedianDuration
-            if (canExceedDeadline) {
-              speculated = checkAndSubmitSpeculatableTask(tid, time, 0)
-            }
-          }
-        }
-        foundTasks |= speculated
+    val timeMs = clock.getTimeMillis()
+    if (numSuccessfulTasks >= minFinishedForSpeculation || numTasks == 1) {
+      val threshold = if (numSuccessfulTasks <= 0 && isSpeculationThresholdSpecified) {
+        speculationTaskDurationThresOpt.get
+      } else {
+        val medianDuration = successfulTaskDurations.median
+        speculationMultiplier * medianDuration
       }
-    } else if (speculationTaskDurationThresOpt.isDefined && speculationTasksLessEqToSlots) {
-      val time = clock.getTimeMillis()
-      val threshold = speculationTaskDurationThresOpt.get
+      val newThreshold = max(threshold, minTimeToSpeculation)
+        // bound based on that.
+      logDebug("Task length threshold for speculation: " + newThreshold)
+      foundTasks = checkAndSubmitSpeculatableTasks(timeMs, newThreshold, numSuccessfulTasks)
+    } else if (isSpeculationThresholdSpecified && speculationTasksLessEqToSlots) {
+      val threshold = max(speculationTaskDurationThresOpt.get, minTimeToSpeculation)
       logDebug(s"Tasks taking longer time than provided speculation threshold: $threshold")
-      for (tid <- runningTasksSet) {
-        foundTasks |= checkAndSubmitSpeculatableTask(tid, time, threshold)
+      foundTasks = checkAndSubmitSpeculatableTasks(timeMs, threshold, numSuccessfulTasks,
+        customizedThreshold = true)
+    }
+    // avoid more warning logs.
+    if (foundTasks) {
+      val elapsedMs = clock.getTimeMillis() - timeMs
+      if (elapsedMs > minTimeToSpeculation) {
+        logWarning(s"Time to checkSpeculatableTasks ${elapsedMs}ms > ${minTimeToSpeculation}ms")
       }
```

Comment:
One of the things which is slightly difficult to reason about in the PR is the change in semantics around `numTasks == 1`. Can we preserve the existing behavior w.r.t `numTasks == 1` ? We can expand the functionality related to it in a follow up PR: where we focus only on that change in behavior.

That is, we remove all the special casing for `numTasks == 1` introduced in this change.